packages/driver/src/cy/commands/sessions.ts

```diff
@@ -0,0 +1,822 @@
+      })
+    },
+
+    async setStorage (options: any, clearAll = false) {
+      const current_origin = $Location.create(window.location.href).origin as string
+
+      const mapToCurrentOrigin = (v) => ({ ...v, origin: (v.origin && v.origin !== 'current_origin') ? $Location.create(v.origin).origin : current_origin })
+
+      const mappedLocalStorage = _.map(options.localStorage, (v) => {
+        const mapped = { origin: v.origin, localStorage: _.pick(v, 'value', 'clear') }
+
+        if (clearAll) {
+          mapped.localStorage.clear = true
+        }
+
+        return mapped
+      }).map(mapToCurrentOrigin)
+
+      const mappedSessionStorage = _.map(options.sessionStorage, (v) => {
+        const mapped = { origin: v.origin, sessionStorage: _.pick(v, 'value', 'clear') }
+
+        if (clearAll) {
+          mapped.sessionStorage.clear = true
+        }
+
+        return mapped
+      }).map(mapToCurrentOrigin)
+
+      const storageOptions = _.map(_.groupBy(mappedLocalStorage.concat(mappedSessionStorage), 'origin'), (v) => _.merge({}, ...v))
+
+      await _setStorageOnOrigins(storageOptions)
+    },
+
+    registerSessionHooks () {
+      Cypress.on('test:before:run:async', () => {
+        if (Cypress.config('experimentalSessionSupport')) {
+          currentTestRegisteredSessions.clear()
+
+          return navigateAboutBlank(false)
+          .then(() => sessions.clearCurrentSessionData())
+          .then(() => {
+            return Cypress.backend('reset:renderedHTMLOrigins')
+          })
+        }
+
+        return
+      })
+    },
+
+  }
+
+  Cypress.on('run:start', () => {
+    sessions.registerSessionHooks()
+  })
+
+  Commands.addAll({
+    session (id, setup?: Function, options: {
+      validate?: Function
+    } = {}) {
+      throwIfNoSessionSupport()
+
+      if (!id || !_.isString(id) && !_.isObject(id)) {
+        throw new Error('cy.session requires a string or object as the first argument')
+      }
+
+      // backup session command so we can set it as codeFrame location for errors later on
+      const sessionCommand = cy.state('current')
+
+      // stringfy determinitically if we were given an object
+      id = typeof id === 'string' ? id : stringifyStable(id)
+
+      if (options) {
+        if (!_.isObject(options)) {
+          throw new Error('cy.session optional third argument must be an object')
+        }
+
+        const validopts = {
+          'validate': 'function',
+        }
+
+        Object.keys(options).forEach((key) => {
+          const expectedType = validopts[key]
+
+          if (!expectedType) {
+            throw new Error(`unexpected option **${key}** passed to cy.session options`)
+          }
+
+          const actualType = typeof options[key]
+
+          if (actualType !== expectedType) {
+            throw new Error(`invalid option **${key}** passed to cy.session options. Expected **${expectedType}**, got ${actualType}`)
+          }
+        })
+      }
```

Comment:
These error messages should be in [error_messages.js](https://github.com/cypress-io/cypress/blob/8550842e3696095c9391e89f30c87af6369ecfb9/packages/driver/src/cypress/error_messages.js), utilizing `cmd('sessions')` so it's formatted consistently. They should all include the value of the argument when it's incorrect. For example, `cy.session() optional third argument must be an object, you passed: true`.

Reply:
85d85eaeeb